testsuite/oauth/authorize.py

```diff
@@ -0,0 +1,236 @@
+""" Authorize the SMART API.
+"""
+from urllib import parse
+import uuid
+
+from pyvirtualdisplay import Display
+from selenium import webdriver
+from selenium.common.exceptions import NoSuchElementException
+from selenium.webdriver.chrome.options import Options
+from selenium.webdriver.support.ui import WebDriverWait
+
+
+AUTHORIZE_TIMEOUT = 15
+
+
+class Authorizer(object):
+    """ Orchestrate the authorization path.
+
+    Attributes:
+        config (dict): The oauth config for this vendor.
+        authorize_url (string): The vendor's authorize endpoint.
+
+    Example:
+        Implements the context manager methods.
+
+            authorizer = Authorizer()
+            with authorizer as browser:
+                token = authorizer.authorize()
+
+        Is equivalent to:
+
+            authorizer = Authorizer()
+            try:
+                authorizer.open()
+                token = authorizer.authorize()
+            finally:
+                authorizer.close()
+    """
+    browser = None
+    config = None
+    authorize_url = None
+    display = None
+    _state = None
+
+    def __init__(self, config, authorize_url):
+        self.config = config
+        self.authorize_url = authorize_url
+
+    def authorize(self):
+        """ The actual authorization method.
+        """
+        if self.browser is None:
+            raise Exception('Webdriver must be connected first.')
+
+        parameters = self.launch_params
+
+        self.ask_for_authorization(parameters)
+        response = self.provide_user_input()
+
+        try:
+            self._validate_state(response)
+            self._validate_code(response)
+        except AssertionError as err:
+            raise ValidationErrorException(str(err), self.browser)
+
+        return response['code'][0]
+
+    def ask_for_authorization(self, parameters):
```

Comment:
Since this isn't a "skeleton" anymore, I renamed these methods to match the language on http://docs.smarthealthit.org/authorization/ a bit better.